starlette/middleware/base.py

```diff
@@ -42,18 +48,42 @@ async def coro() -> None:
         task = loop.create_task(coro())
         message = await queue.get()
         if message is None:
+            queue.task_done()
             task.result()
             raise RuntimeError("No response returned.")
         assert message["type"] == "http.response.start"
 
         async def body_stream() -> typing.AsyncGenerator[bytes, None]:
-            while True:
+            def streaming_predicate(
+                msg: typing.Optional[dict], more_body: bool = True
+            ) -> bool:
+                return (
+                    msg is not None
+                    and msg["type"] == "http.response.body"
+                    and "more_body" in msg
+                    and msg["more_body"] is more_body
+                )
+
+            # In non-streaming responses, there will be one message to emit
+            message = await queue.get()
+            if message is not None:
+                queue.task_done()
+                assert message["type"] == "http.response.body"
+                yield message.get("body", b"")
+
+            while streaming_predicate(message, more_body=True):
                 message = await queue.get()
+                queue.task_done()
                 if message is None:
                     break
                 assert message["type"] == "http.response.body"
                 yield message.get("body", b"")
-            task.result()
+
+            try:
+                task.result()  # check for exceptions and raise if present
```

Comment:
Alternatively:
```python
if task.done():
    task.result()
```
I guess it's more Pythonic to "ask for forgiveness rather than permission" so I'm ok with this, but IMO `if task.done()` is more readable.

Reply:
I think that is an improvement and I agree it is more readable. Thanks for the suggestion. I always forget that `done` will be `True` even if an exception is raised (in my mind `done` means "finished without throwing an exception.")